reference/src/layout/arrays.md

```diff
@@ -0,0 +1,33 @@
+# Layout of Rust array types
+
+Array types, `[T; N]`, store `N` values of type `T` contiguously with a constant
+_stride_ (the distance between each two consecutive values).
+
+The offset of the first array element is `0`.
+
+The stride of the array is computed as the size of the element type rounded up
+to the next multiple of the alignment of the element type. That is, the _stride_
+of an array can be larger than the element size iff the alignment requirements
+of the element are larger than its size.
+
+Having a stride larger than the element size allows:
+
+```rust,ignore
+struct A(u16, u8); // size_of == 3, align_of == 4
+type B = [A; 4];  // => stride == 4 > 3
+```
+
+The size and alignment of `Vector` element types match, such that `Vector` types
+and arrays are layout compatible.
+
+`repr(C)` arrays have the same layout as C arrays and are passed by pointer in C
```

Comment:
There's no such thing as a "`repr(C)` array". Do you mean an array those *element type* is compatible with some C type?

Reply:
I think I meant just "arrays", as in "arrays are passed by pointer in C FFI" (nothing else).  cc @RalfJung 

What I was referring to is that in: 

```rust
extern "C" { fn foo(x: [i32; 2]); }
unsafe { foo([0; 2]) }; 
```

the array is not memcpy'ed into `foo`, but passed by pointer. 

But maybe it is not worth to mention this here; that's ABI / calling convention and not layout.

Reply:
No that's not true. The array is passed as an aggregate, which depending on ABI details might end up on the stack, but it usually won't be ABI compatible with a C function that looks like it takes an array argument. [See LLVM IR here, for example.](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=8d07d1734f3b2ff396dbe859aca5887c)

Reply:
Yeah... I don't think fixed-size arrays are FFI-safe. Does the lint not fire for them?

Reply:
Indeed. Arrays are Aggregates at least in the SysV ABI, so that makes sense.

Reply:
> Does the lint not fire for them?

No, it does not. I've removed this.

The only open question is whether this should mention that arrays are not FFI safe.

Reply:
Well, Rust arrays are definitely layout compatible with C arrays, and if they're contained in a struct the struct will also be ABI compatible. The root of the issue is that a C signature like `void foo(int x[4]);` does not actually mean that an array is passed, it decays to a pointer. This is definitely worth noting (and IMO the ctypes lint should fire for array arguments) but it's actually "just" a subtlety of C surface syntax, not an actual difference in ABI rules.

Reply:
So I worded the issues with C FFI in the last commit. I hope this is now ok.